### PR TITLE
Added a Benefits of Using the Overlay Graph Driver with SELinux section

### PR DESCRIPTION
**Pull request by ahardin-rh**
https://trello.com/c/C1QKofOk/478-3-scale37-ocp-document-openshift-with-docker-overlay2selinux-enabled-performance-improvements

@jeremyeder Here's my first pass at expanding the documentation on Overlay + SELInux. PTAL. Thanks!

### CODE REVIEW

**Comment by jeremyeder** [comment]
Thanks Ashley -- I'll need to tweak some of it. Will put some comments in soon.

**Comment by ahardin-rh** [comment]
@jeremyeder bump

File: scaling_performance/optimizing_storage.adoc [hunk 1]
```diff
@@ -126,5 +126,21 @@ Overlay is also supported for Docker use cases as of Red Hat Enterprise Linux
 7.2, and provides faster start up time and page cache sharing, which can
 potentially improve density by reducing overall memory utilization.
 
+[[benefits-of-using-the-overlay-graph-driver]]
+== Benefits of Using the Overlay Graph Driver with SELinux
 
+Now that Overlay is in the upstream kernel and support for SELinux is extended,
```
Comment (jeremyeder) [inline scaling_performance/optimizing_storage.adoc]: This paragraph needs to be largely re-written, I will give it a shot.

File: scaling_performance/optimizing_storage.adoc [hunk 2]
```diff
@@ -126,5 +126,21 @@ Overlay is also supported for Docker use cases as of Red Hat Enterprise Linux
...
+Overlay file system has several advantages over device mapper (most importantly,
+page cache sharing among containers). Support for the Overlay file system was
+added to Red Hat Enterprise Linux with salient caveats: It was not POSIX
+compliant and the use of Overlay was, at the time, incompatible with SELinux (a
+key security and isolation technology). The density improvements gained by page
+cache sharing are very important for certain environments where there is
+significant overlap in base image content.
+
+Red Hat testing found Overlay2 to be very stable, and that stability is
+furthered with the addition of SELinux support.
```
Comment (jeremyeder) [inline scaling_performance/optimizing_storage.adoc]: How about this:

The default docker storage configuration on RHEL remains device-mapper.  As of Red Hat Enterprise Linux 7.2, overlay became a supported graph driver.  And, as of Red Hat Enterprise Linux 7.4, SELinux and the overlay2 graph driver became a supported combination.

The main advantage of the overlay filesystem is Linux page cache sharing among containers on the same node.  This attribute of overlay leads to reduced I/O during container startup (and thus faster container startup time by several hundred milliseconds), as well as reduced memory usage when similar images are running on a node.  Both side-effects lend themselves well towards many environments, especially those with the goal of optimizing for density and/or that have high container churn rate (such as a build farm), or those that have significant overlap in image content.

Page cache sharing is not possible with device-mapper due to how allocating thin-provisioned devices on a per-container basis.

**Review by jeremyeder — CHANGES_REQUESTED** [review]

**Comment by ahardin-rh** [comment]
@jeremyeder Updated. Thanks!

@openshift/team-documentation Please peer review

**Comment by jeremyeder** [comment]
@rhatdan @rhvgoyal can you please review this small doc paragraph?

**Comment by rhatdan** [comment]
LGTM  (Only question I would have as a customer would why is not this the default?)

**Comment by jeremyeder** [comment]
@rhatdan Thanks!  I didn't want to speculate, and I didn't feel mentioning Fedora 27's change to Overlay as default made sense in enterprise docs because it does not imply any *guarantee* that RHEL will change.  If you feel otherwise, can you propose a sentence or two to add to these docs?

File: scaling_performance/optimizing_storage.adoc [hunk 3]
```diff
@@ -126,5 +126,22 @@ Overlay is also supported for Docker use cases as of Red Hat Enterprise Linux
...
+[[benefits-of-using-the-overlay-graph-driver]]
+== Benefits of Using the Overlay Graph Driver with SELinux
+
+The default Docker storage configuration on Red Hat Enterprise Linux (RHEL)
+continues to be device-mapper. As of RHEL 7.2, Overlay became a supported graph
+driver. As of RHEL 7.4, SELinux and the Overlay2 graph driver became a supported
+combination.
+
+The main advantage of the Overlay file system is Linux page cache sharing among
+containers on the same node. This attribute of Overlay leads to reduced
```
Comment (rhvgoyal) [inline scaling_performance/optimizing_storage.adoc]: I would say mention page cache sharing happens only if containers are sharing any image. so something like, `page cache sharing among containers sharing image on same node`

File: scaling_performance/optimizing_storage.adoc [hunk 4]
```diff
@@ -21,11 +21,11 @@ limitations of a union file system, and does not yet support SELinux.
 
 For more information about Overlay, including supportability and usage caveats,
 link:https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html-single/7.3_Release_Notes/index.html#technology_previews_file_systems[see
-the RHEL 7.3 Release Notes].
+the Red Hat Enterprise Linux (RHEL) 7 Release Notes].
 
 In production environments, using a LVM thin pool on top of regular block
 devices (not loop devices) for container images and container root file systems
-storage is recommended. 
+storage is recommended.
```
Comment (mburke5678) [inline scaling_performance/optimizing_storage.adoc]: Is "root file systems storage" correct? Just sounds odd.
"root file system storage"?

File: scaling_performance/optimizing_storage.adoc [hunk 5]
```diff
@@ -126,5 +126,22 @@ Overlay is also supported for Docker use cases as of Red Hat Enterprise Linux
 7.2, and provides faster start up time and page cache sharing, which can
 potentially improve density by reducing overall memory utilization.
 
-
-
+[[benefits-of-using-the-overlay-graph-driver]]
+== Benefits of Using the Overlay Graph Driver with SELinux
+
+The default Docker storage configuration on Red Hat Enterprise Linux (RHEL)
+continues to be device-mapper. As of RHEL 7.2, Overlay became a supported graph
```
Comment (mburke5678) [inline scaling_performance/optimizing_storage.adoc]: "continues to be a/the device-mapper"?
Also, do we need the hyphen in device-mapper?  Super quick Google search shows no hyphen.

File: scaling_performance/optimizing_storage.adoc [hunk 6]
```diff
@@ -126,5 +126,22 @@ Overlay is also supported for Docker use cases as of Red Hat Enterprise Linux
...
+The main advantage of the Overlay file system is Linux page cache sharing among
+containers on the same node. This attribute of Overlay leads to reduced
+input/output (I/O) during container startup (and, thus, faster container startup
+time by several hundred milliseconds), as well as reduced memory usage when
+similar images are running on a node. Both side-effects lend themselves well to
+many environments, especially those with the goal of optimizing for density and
+have high container churn rate (such as a build farm), or those that have
+significant overlap in image content.
+
+Page cache sharing is not possible with device-mapper because thin-provisioned
```
Comment (mburke5678) [inline scaling_performance/optimizing_storage.adoc]: Same here "device-mapper"

File: scaling_performance/optimizing_storage.adoc [hunk 7]
```diff
@@ -126,5 +126,22 @@ Overlay is also supported for Docker use cases as of Red Hat Enterprise Linux
...
+The default Docker storage configuration on Red Hat Enterprise Linux (RHEL)
+continues to be device-mapper. As of RHEL 7.2, Overlay became a supported graph
+driver. As of RHEL 7.4, SELinux and the Overlay2 graph driver became a supported
+combination.
+
+The main advantage of the Overlay file system is Linux page cache sharing among
+containers on the same node. This attribute of Overlay leads to reduced
+input/output (I/O) during container startup (and, thus, faster container startup
+time by several hundred milliseconds), as well as reduced memory usage when
+similar images are running on a node. Both side-effects lend themselves well to
```
Comment (mburke5678) [inline scaling_performance/optimizing_storage.adoc]: "Both side-effects lend themselves well" seems too marketing-speak to me. Not sure what to replace it with. "These effects are beneficial in many environments..." ??

**Comment by mburke5678** [comment]
@ahardin-rh Couple of comments.

**Comment by ahardin-rh** [comment]
Thanks, all! This is updated.

**Comment by rhatdan** [comment]
@jeremyeder Perhaps.

As we monitor the use of Overlay as the containers storage backend,  we will consider moving RHEL to Overlay as default in future releases.

**Comment by ahardin-rh** [comment]
@jeremyeder @rhatdan Thanks! Updated.

**Comment by jeremyeder** [comment]
@ahardin-rh I don't see @rhatdan sentence in the PR?

**Comment by ahardin-rh** [comment]
@jeremyeder @rhatdan I edited it a bit to avoid using "we": 
While the use of Overlay as the containers storage back-end is being monitored, moving RHEL to Overlay as the default in future releases is under consideration.

**Comment by jeremyeder** [comment]
LGTM

**Comment by rhatdan** [comment]
LGTM

**Comment by ahardin-rh** [comment]
[rev_history]
|xref:../scaling_performance/optimizing_storage.adoc#scaling-performance-optimizing-storage[Optimizing Storage]
|Added a xref:../scaling_performance/optimizing_storage.adoc#benefits-of-using-the-overlay-graph-driver[Benefits of Using the Overlay Graph Driver with SELinux] section.
%

**Pull request merged**